docs/guides/ssr.md

```diff
@@ -111,6 +180,95 @@ function Posts() {
 
 As demonstrated, it's fine to prefetch some queries and let others fetch on the queryClient. This means you can control what content server renders or not by adding or removing `prefetchQuery` for a specific query.
 
+#### With experimental `app` directory
+Wrap the `<QueryClientProvier>` and `<Hydrate>` components inside a Client Component that initializes the `QueryClient` your app will use on the client.
+
+- `QueryClient` should be created on an instance ref (or in React State ). **This ensures that data is not shared between different users and requests, while still only creating the QueryClient once per component lifecycle.**
+- The `dehydratedState` prop will contain your prefetched data along with `dataUpdatedAt` which provides the time the data was fetched on the server
+
+```tsx
+// app/react-query.jsx
+'use client'
+
+import { QueryClient, QueryClientProvider, Hydrate } from '@tanstack/react-query'
+
+export default function ReactQuery({dehydratedState, children}) {
+  const [queryClient] = React.useState(() => new QueryClient())
+  
+  return (
+    <QueryClientProvider client={queryClient}>
+      <Hydrate state={dehydratedState}>
+        {children}
+      </Hydrate>
+    </QueryClientProvider>
+  )
+}
+```
+
+Prefetch your data inside a Server Component and provide your previously defined Client Component with the dehydrated state of the server-side `QueryClient`.
+
+- Create a new QueryClient instance **for each page request. This ensures that data is not shared between users and requests.**
+- Prefetch the data using the clients prefetchQuery method and wait for it to complete
+- Use dehydrate to dehydrate the query cache and pass it to your wrapped client provider via the dehydratedState prop
+- Wrap your component tree with this new Server Component provider
+
+```tsx
+// app/providers.jsx
+import { QueryClient, dehydrate } from '@tanstack/query-core'
+import ReactQuery from './react-query'
+
+export default async function Providers({children}) {
+  const queryClient = new QueryClient()
+  await queryClient.prefetchQuery(['posts'], getPosts)
```

Comment:
Seems really backwards that you gotta `await queryClient.prefetchQuery(['posts'], getPosts)` in order for the state to be hydrated... 

I would expect some magic to hydrate anything that has been called with `useQuery(..., { suspense: true })` automatically

Reply:
Is there a missing API in React for this at the moment? No way of pushing state from a server-render React-component to a client-side one without this stuff?

Reply:
Server-side suspense is a whole other wrinkle I hadn't considered. Dropping https://github.com/reactwg/react-18/discussions/37 here for reference because I'll likely want to circle back on this.

Reply:
I agree that this could be handled better by TanStack/query, but for now, it is what it is :). This is exactly why I'd like the whole section to be marked as `experimental` - because the recommendations might change.

also looping in @Ephem :)

Reply:
I mean, the recommendation should always be to prefetch things as early as possible, both to get a headstart and to avoid waterfalls where you suspend in A to be able to render B which then suspend again etc. We should definitely investigate supporting the other case as well though (if it's even going to be possible), but that's for the future.

One caveat is that currently you _have_ to await the data in the RSC, which blocks streaming what you can early. What we really want is to _start_ prefetching in the RSC but not await it, instead suspending in the component that needs this data. That will require both that React implements de/hydration of Promises and a bunch of changes to RQ though, so I think this PR (with the proposed changes elsewhere) documents the best current experimental approach.

Reply:
Oh, and like we talked about on Discord @KATT, I think it's interesting here to think about how to support "normal" streaming rendering without RSCs as well. A lot of frameworks might implement streaming rendering without RSCs, so that's also something we want to support.

Reply:
Btw, for now we should definitely document that `useQuery(..., { suspense: true })` wont work on the server unless you have already prefetched that data.

Reply:
If you've already prefetched your data, is there any utility to set a client-side suspense boundary? If you don't prefetch, you can set suspense boundaries inside a client-side component and use `useQuery(..., { suspense: true })`. Data will start fetching after the page loads. If you're prefetching, you can set a suspense boundary in a server-side component, and there's no need for `{ suspense: true }` inside `useQuery`. Data will start fetching as soon as possible, the server will send a response without waiting for it, and then stream the rendered component to be integrated into the page once the fetching completes.